File: testers/sqlserver/java/src/java/testdb/SqlServerTestHelper.java
package testdb;

import dev.typr.foundations.Connection;
import dev.typr.foundations.ConnectionRead;
import dev.typr.foundations.SqlConsumer;
import dev.typr.foundations.SqlFunction;
import dev.typr.foundations.Transactor;
import dev.typr.foundations.connect.ConnectionSettings;
import dev.typr.foundations.connect.SqlServerConfig;
import dev.typr.foundations.connect.SqlServerEncrypt;
import dev.typr.foundations.connect.TransactionIsolation;

public class SqlServerTestHelper {
  private static final SqlServerConfig CONFIG =
      SqlServerConfig.builder("localhost", 1433, "typr", "sa", "YourStrong@Passw0rd")
          .encrypt(SqlServerEncrypt.FALSE)
          .build();

  // SQL Server uses pessimistic locking by default, which causes deadlocks when
  // multiple tests run in parallel and access the same tables. READ_UNCOMMITTED
  // prevents lock contention. Since tests rollback anyway, dirty reads are fine.
  private static final ConnectionSettings SETTINGS =
      ConnectionSettings.builder()
          .transactionIsolation(TransactionIsolation.READ_UNCOMMITTED)
          .build();

  private static final Transactor TRANSACTOR = Transactor.create(CONFIG, SETTINGS).rollbackOnly();

  public static <T> T apply(SqlFunction<Connection, T> f) {
    return TRANSACTOR.transact(f);
  }

  public static void run(SqlConsumer<Connection> f) {
    TRANSACTOR.transactVoid(f);
  }

  public static <T> T applyRead(SqlFunction<ConnectionRead, T> f) {
    return TRANSACTOR.transactRead(f);
  }
}

File: testers/sqlserver/kotlin/src/kotlin/testdb/SqlServerTestHelper.kt
package testdb

import dev.typr.foundationskt.Connection
import dev.typr.foundationskt.ConnectionRead
import dev.typr.foundationskt.Transactor
import dev.typr.foundationskt.connect.ConnectionSettings
import dev.typr.foundationskt.connect.SqlServerConfig
import dev.typr.foundationskt.connect.SqlServerEncrypt
import dev.typr.foundationskt.connect.TransactionIsolation

object SqlServerTestHelper {
    private val CONFIG: SqlServerConfig =
        SqlServerConfig.builder("localhost", 1433, "typr", "sa", "YourStrong@Passw0rd")
            .encrypt(SqlServerEncrypt.FALSE)
            .build()

    private val SETTINGS: ConnectionSettings =
        ConnectionSettings.builder()
            .transactionIsolation(TransactionIsolation.READ_UNCOMMITTED)
            .build()

    private val TRANSACTOR: Transactor = Transactor.create(CONFIG, SETTINGS).rollbackOnly()

    fun <T> apply(f: (Connection) -> T): T = TRANSACTOR.transact(f)

    fun run(f: (Connection) -> Unit) = TRANSACTOR.transact { conn -> f(conn) }

    fun <T> applyRead(f: (ConnectionRead) -> T): T = TRANSACTOR.transactRead(f)
}
